New definition for "prefixes of words"

avekens opened the pull request:
As discussed in #1648, the new definition for "prefixes of words" df-pfx is moved from my mathbox to main set.mm.

Until now:
* df-pfx and related theorems are moved from AV's mathbox to main
* suffix OLD is appended to theorems which are replaced by theorems using the prefix operator.
* theorems are rearranged, mainly to have OLD theorems near to the current ones
* revision of comments:
** "Could replace.." replaced by original contribution tag.
** OLD theorems marked as obsolete
* OLD theorems are replaced in the proofs of the following theorems: ~psgnunilem5 ~wrdind ~wrd2ind ~splval2 ~cshwidxmod ~wrdeqs1cat ~swrdccat3b ~swrdccat ~ccatopth ~splid ~cshw0 ~iwrdsplit (as a side effect, this saved 324 bytes in total)

TODO:
* revision of df-splice 
* replace remaining OLD theorems used in proofs (mainly of theorems of Graph Theory) - I assume that this will shorten again the proofs, hopefully significantly.
* rewrap, especially the proofs of the OLD theorems (I didn't it yet intentionally, to make reviews easier).

This is an intermediate PR not intended to be merged immediately. The TODOs listed above should be done first, but I want to publish the material already now to get an early feedback (is somebody is interested).

benjub commented on set.mm:
```diff
@@ -147746,6 +147746,473 @@ of an open range of nonnegative integers (of length equal to the length of
+  abbreviation for "prefix".  Regarding the meaning of "prefix", it is
+  different in computer science (automata theory/formal languages) compared
+  with linguistics: in linguistics, a prefix has a meaning (see Wikipedia
+  "Prefix" ~ https://en.wikipedia.org/wiki/Prefix ), whereas in computer
```
No need for these five lines, IMO.

avekens replied:
Which five lines exactly? The section comment consists of 10 lines...

benjub replied:
The five quoted lines. That is, you can begin at "A prefix is an arbitrary..."

avekens replied:
OK. I have restructured this text a little bit, because I would like to keep this background information.

avekens replied:
The whole section header comment is removed now. Some of its contents is moved to the comment of df-pfx.

benjub commented on set.mm:
```diff
@@ -147746,6 +147746,473 @@ of an open range of nonnegative integers (of length equal to the length of
+  ${
+    $d l s $.
+    $( Define an operation which extracts prefixes of words, i.e. subwords
+       starting at the beginning of a word.  Definition in section 9.1 of
```
There is no definition number ?
Capitalize "Section"

avekens replied:
> There is no definition number ? 

Unfortunately, there is no definition number.

>Capitalize "Section"

OK

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147746,6 +147746,473 @@ of an open range of nonnegative integers (of length equal to the length of
+      BIGDJZKEJZLZMNZAKBLZMNZOIODEIGUOTPUKDEQRULAPZUMBPZHZUOUQPUKUTULAUNUPMURUS
+      UAUSUNUPPURUMBKUBSUCSUIAIFUJACUDUEUIUJUFUKAUPMUGUH $.
+
+    $( A zero length prefix.  (Contributed by AV, 2-May-2020.) $)
```
The zero length prefix is the empty set.

avekens replied:
Done

benjub commented:
I began, but it's rather long. Is it better to review commits independently or grouped ?

avekens commented:
> I began, but it's rather long. Is it better to review commits independently or grouped ?

Until now, most of the intellectual work was already done in my mathbox. The current two commits are almost only editorial work.

In the first commit, df-pfx and related theorems are moved from my mathbox to main and suffix OLD is appended to theorems which are replaced by theorems using the prefix operator. These are mainly theorems containing the pattern ( * substr <. 0 , * >. ).

In the second commit, the theorems are rearranged a little bit, and the comments are revised. Only the replacenment of OLD theorems in the proofs of the theorems mentioned in the description of this PR is of substantial interest: ~psgnunilem5 ~wrdind ~wrd2ind ~splval2 ~cshwidxmod ~wrdeqs1cat ~swrdccat3b ~swrdccat ~ccatopth ~splid ~cshw0 ~iwrdsplit : but except checking that most of their proofs are shortened nothing meaningful can be done by a review.

Therefore, I think it makes no great difference if the first commit or the whole PR is reviewed.

Maybe my next commit (revision of the splice operator) will be more interesting...

benjub commented:
Ok, so as I understand, there will be a next commit, and also somes checks did not pass. So I'll wait for your next commit(s). To indicate explicitly that more is coming, you can "convert to draft" in the upper-right corner, and then convert back when it's ready for review.

avekens commented:
> Ok, so as I understand, there will be a next commit, and also somes checks did not pass. So I'll wait for your next commit(s). To indicate explicitly that more is coming, you can "convert to draft" in the upper-right corner, and then convert back when it's ready for review.

Sorry, the intention of this (draft) PR was to get an early feadback, not a thorough review. 

I will finish the next step (revision of definition of `splice`) soon, provide the changes in a separate commit, and then clean up the material for a proper PR. The remaining tasks - replace remaining OLD theorems used in proofs (mainly of theorems of Graph Theory) - will be done in a separate PR.

avekens commented:
The following proofs were shortened using the prefix operator, many of them significantly (and only three proofs became larger). It is expected that many additional proofs (which still used the OLD theorems) can be shortened, too.

- splval (687 bytes -> 657 bytes) -30
- splcl (not used) (694 bytes -> 671 bytes) -23
- splid (not used) (569 bytes -> 450 bytes) -119
- spllen (not used) (996 bytes -> 940 bytes) -56
- splfv1 (not used) (1058 bytes -> 850 bytes) -208
- splfv2a (not used) (1142 bytes -> 1064 bytes) -78
- splval2 (not used) (1510 bytes -> 1447 bytes) -63
- -----
- psgnunilem5 (3324 bytes-> 3092 bytes) shortened! -232
- wrdind (1982 bytes -> 1957 bytes) shortened! -25
- wrd2ind (5134 bytes -> 5081 bytes) shortened! -53
- cshwidxmod (3640 bytes -> 3594 bytes) -46
- wrdeqs1cat (625 bytes -> 387 bytes) -238
- swrdccat3b (841 bytes -> 825 bytes) -16
- swrdccat (2499 bytes -> 2880 bytes) + 381
- ccatopth (651 bytes -> 626 bytes) -25
- cshw0 (582 bytes -> 626 bytes) +44
- iwrdsplit (908 bytes -> 839 bytes) -69
- -----
- gsumspl ( (721 bytes -> 723 bytes) +2
- efgredleme (7237 bytes -> 7066 bytes) -271
- efgredlemd (4579 bytes -> 4489 bytes) -90
- efgredlemc (5900 bytes -> 5813 bytes) -87
- efgcpbllemb (5393 bytes -> 5317 bytes) -76
- frgpuplem (4565 bytes -> 4463 bytes) -102

In total, 1480 bytes were saved.

avekens commented:
This PR is teady for review. As I mentioned in #1648, the remaining OLD theorems used in proofs will be replaced after this PR is accepted and merged.

benjub commented:
> This PR is teady for review. As I mentioned in #1648, the remaining OLD theorems used in proofs will be replaced after this PR is accepted and merged.

I will. Hopefully tomorrow.

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+  ~ https://math.stackexchange.com/questions/2190559/
+  is-there-standard-terminology-notation-for-the-prefix-of-a-word ).
+
+  This is the meaning of "prefix" in computer science (automata theory or
+  formal languages) , whereas in linguistics, a prefix has a meaning (see
+  Wikipedia "Prefix" ~ https://en.wikipedia.org/wiki/Prefix ).
```
If you want. But that's the same with every word: "ball", "union"... do not have the same meaning in everyday language.

avekens replied:
I added it (and would like to keep it) mainly because the Wikipedia page "prefix" is about the linguistic meaning...

benjub replied:
Ok. Then, "has a DIFFERENT meaning"

avekens replied:
The whole section header comment is removed now.

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+    $d l s $.
+    $( Define an operation which extracts prefixes of words, i.e. subwords
+       starting at the beginning of a word.  Definition in Section 9.1 of
+       [AhoHopUll] p. 318.  "pfx" is used as label fragment (as propose in
+       ~ https://www.allacronyms.com/prefix/abbreviated .  (Contributed by AV,
```
"proposeD"
But I think you can remove that parenthesis anyway (there is no closing parenthesis, by the way). There is no need to justify our label choices by websites, especially nonmathematical ones.

avekens replied:
> "proposeD" But I think you can remove that parenthesis anyway (there is no closing parenthesis, by the way). 

Done

>There is no need to justify our label choices by websites, especially nonmathematical ones.

I added this because there was a long discussion about the label in #1648.

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+      UCULUPUJCGUKUDUECGUKUFUGUHUI $.
+
+    $d V x $.  $d W x $.
+    $( A prefix of a word is a function from a half-open range of nonnegative
```
Is "half-open" the terminology used in set.mm ? (because for integer intervals, there is no notion of open or closed, but maybe it was decided in the past to mean "upper bound excluded" ?) So do as you see fit.

avekens replied:
Yes, this is the offical wording used in set.mm, see "5.5.7  Half-open integer ranges".

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+    VEVAAPVCUFMEZVIVAVBVDUGVEATEZVCTEZVDVLVBVAVMVDAUHRVAVBVNVDBCUIUJVAVBVDUKAVC
+    UMULABCUNUOVIVKVJVFVGBUPUQURUSUT $.
+
+  $( The value of the prefix extractor is the empty set (undefined) if the
+     argument is not within the range of the word.  (Contributed by AV,
```
Maybe I am not using the correct words, but I'd prefer something like:
> The value of a prefix operation when the integer argument is larger than the word length is the empty set. (This is due to our definition of function values for out-of-domain arguments, see ~iotanul ).

It was probably discussed already, but: have you thought about the definition which would give the whole word in that case (and more generally, similarly for substrings) ?

avekens replied:
Comment revised.

> It was probably discussed already, but: have you thought about the definition which would give the whole word in that case (and more generally, similarly for substrings) ?

No, I did not consider this, and I don't know if this was discussed by others. I cannot see any advantage of this at the moment, and the definition(s) would become wore complicated.

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+
+  $( Length of a prefix of a word reduced by a single symbol.  (Contributed by
+     AV, 4-Aug-2018.)  (Revised by AV, 3-May-2020.) $)
+  pfxlen0 $p |- ( ( W e. Word V /\ L e. NN0 /\ ( # ` W ) = ( L + 1 ) )
+                  -> ( # ` ( W prefix L ) ) = L ) $=
```
isn't it enough to suppose `L <= length(W)` ?

avekens replied:
> isn't it enough to suppose `L <= length(W)` ?

This is covered by ~ pfxlen.

But you are right, ~pfxlen0 makes no sense, and its predecessor swrd0len0(OLD) is only used once. Therefor I will delete ~pfxlen0.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+    ZIJEZKZCALJGHZCAUIMNJGHZOJAUIARJZOJZUIUKULAUMUNOBCAPABCQUAUJASEUISEUOUIUBUH
+    UJAAUIUCTUHUIBCUDTAUIUEUFUG $.
+
+  $( The first symbol in a prefix of a word.  (Contributed by Alexander van der
```
Maybe
> The first symbol of a prefix is the first symbol of the word.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+    DZMBVCRPGMBGSUSVAUAABUBVBVCUCDZVDVBUTEUDGDZVEVBETDZUTTDZVAVFVGVBUEUFUSVHVAU
+    SUTABUGUHUIUSVAUJEUTUKULUTUMUNVCUOUPMABUQUR $.
+
+  $( The last symbol in a (not empty) prefix of a word.  (Contributed by
```
in a nonempty prefix

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,405 @@ of an open range of nonnegative integers (of length equal to the length of
+    UN $.
+
+  ${
+    $d M i $.  $d N i $.  $d U i $.  $d V i $.  $d W i $.
+    $( The prefixes of two words are equal iff they have the same length and
+       the same symbols at each position.  (Contributed by Alexander van der
+       Vekens, 7-Aug-2018.)  (Revised by AV, 4-May-2020.) $)
```
That's true for any word. Is it useful to have it for prefixes ?

avekens replied:
This theorem (or its predecessor ~ swrdeqOLD) is used 3 times, therefore I would like to keep it.

benjub commented:
I made a few comments. So it's kind of reviewed up to the latest comment I wrote (Line 147400). I have to say the diff is huge and I do not have much to say in that topic. By the way, why are all these things about words in the part "Real and complex numbers"? Can someone review a bit more ? Maybe someone who took part in #1648 can chime in ?

avekens commented:
> I made a few comments. So it's kind of reviewed up to the latest comment I wrote (Line 147400). 

@benjub thank you very much for your review. I considered/commented your remarks as far as possible.

>I have to say the diff is huge and I do not have much to say in that topic. 

A review is meaningful only up to line 148000 - the rest is almost only technical.

> By the way, why are all these things about words in the part "Real and complex numbers"? 

I do not know why it was placed here (in 2014?), but since the main prerequisites are the integers and the size function, which are also topics of this part and definied immediatly before words, I think it is a good place. If this should be discussed further, this should be done in another place (open an issue for that?). 

> Can someone review a bit more ? Maybe someone who took part in #1648 can chime in ?

@savask @david-a-wheeler do you want to have a (quick) look at this PR?

savask commented:
I looked at the changes. I don't have any serious remarks, besides perhaps the comment for `df-pfx`:
```
    $( Define an operation which extracts prefixes of words, i.e. subwords
       starting at the beginning of a word.  Definition in Section 9.1 of
       [AhoHopUll] p. 318.  "pfx" is used as label fragment as proposed in
       ~ https://www.allacronyms.com/prefix/abbreviated .  (Contributed by AV,
       2-May-2020.) $)
```
In my opinion, the reference to [AhoHopUll] is not required as the definition of a prefix is intuitive enough and, moreover, the definition was given at the start of the section `Prefixes of a word`. If you want to keep the reference it is much better to put it at the beginning of a section (replacing the stackexchange link, for example). I also suggest removing the reference to allacronyms.com completely, as it's not a credible source and we don't usually express any justification for using a specific abbreviation in set.mm.

On the other hand it might be beneficial to hint the way this function works, for instance, like this:
```
Define an operation which extracts prefixes of words, i.e. subwords
starting at the beginning of a word. In other words, ` ( S prefix L ) ` is
a prefix of ` S ` of length ` L `.  (Contributed by AV, 2-May-2020.) 
```

On a more general note, I can't get rid of the feeling that it's a huge bulk of technical work without any immediate merit. Sure some theorems are shorter, but it would be great to see a nontrivial result involving prefixes (something like the "commuting strings theorem", i.e. `xy = yx` implies `x = z^n` and `y = z^m` for some string `z`. The proof uses prefixes, as far as I remember).

benjub commented:
> ` ( S prefix L ) ` is a prefix of ` S ` of length ` L `.

Maybe use the definite article, and add a provision: ` ( S prefix L ) ` is the prefix of ` S ` of length ` L ` (when the natural `L` is not larger than the length of the word `S`).

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,398 @@ of an open range of nonnegative integers (of length equal to the length of
+
+  ${
+    $d N y $.  $d V x $.  $d x y $.
+    $( Sets of words are disjoint if each set contains extensions of distinct
```
contain ONLY

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,398 @@ of an open range of nonnegative integers (of length equal to the length of
+
+  ${
+    $d x A $.  $d x S $.  $d x Y $.  $d x Z $.
+    $( Joining a prefix with an adjacent subword makes a longer prefix.
```
Joining --> Concatenating
(I haven't checked other occurrences, but let's be consistent with "concatenate", "concatenation".)

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,398 @@ of an open range of nonnegative integers (of length equal to the length of
+
+  ${
+    $d k B $.  $d k S $.  $d k T $.
+    $( Recover the left half of a concatenated word.  (Contributed by Mario
```
Maybe:
The left "concatenand" of a concatenated word is the prefix of that word of its own length.

avekens replied:
The comment was written by Mario, I  do not want to change it without his acceptance. In Addition, I think the word "concatenand" is not commonly used.

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,398 @@ of an open range of nonnegative integers (of length equal to the length of
+      VNVOWRWTXALABCWQVJVKTVLT $.
+  $}
+
+  $( A prefix of length 1.  (Contributed by AV, 15-May-2020.) $)
```
Not sure I'm using the correct words, but maybe:
The prefix of length one of a given nonempty word expressed as a single character word.

avekens replied:
Done (The prefix of length one of a nonempty word expressed as a singleton word.)

benjub commented on set.mm:
```diff
@@ -147273,9 +147660,24 @@ of an open range of nonnegative integers (of length equal to the length of
-  $( A subword of a prefix.  (Contributed by Alexander van der Vekens,
-     6-Apr-2018.) $)
-  swrdswrd0 $p |- ( ( W e. Word V /\ N e. ( 0 ... ( # ` W ) ) )
+  $( A prefix of a subword.  (Contributed by AV, 2-Apr-2018.)  (Revised by AV,
```
A prefix of a subword is a subword.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147288,24 +147690,66 @@ of an open range of nonnegative integers (of length equal to the length of
-  $( A prefix of a prefix.  (Contributed by Alexander van der Vekens,
-     7-Apr-2018.) $)
-  swrd0swrd0 $p |- ( ( W e. Word V /\ N e. ( 0 ... ( # ` W ) )
+  $( A subword of a prefix.  (Contributed by Alexander van der Vekens,
```
A subword of a prefix is a subword.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147288,24 +147690,66 @@ of an open range of nonnegative integers (of length equal to the length of
-  $( A prefix of a prefix with the same length is the prefix.  (Contributed by
-     AV, 5-Apr-2018.)  (Proof shortened by AV, 14-Oct-2018.) $)
-  swrd0swrdid $p |- ( ( W e. Word V /\ N e. ( 0 ... ( # ` W ) ) )
+  $( A prefix of a prefix.  (Contributed by Alexander van der Vekens,
```
A prefix of a prefix is a prefix.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147288,24 +147690,66 @@ of an open range of nonnegative integers (of length equal to the length of
+    adantl swrd0swrd0OLD mpd3an3 ) CBDEZAFCGHZIJEZAFAIJEZCFAKZLJZUELJUFMUCUDUAU
+    CANEUDAUBOAPQRAABCST $.
+
+  $( A prefix of a prefix with the same length is the prefix.  (Contributed by
```
is the ORIGINAL prefix. In other words, the operation "prefix of length N" is idempotent.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147314,83 +147758,180 @@ of an open range of nonnegative integers (of length equal to the length of
+    LBQWBABVRPVS $.
+
+  $( The last symbol of a word concatenated with the word with the last symbol
+     removed having results in the word itself.  (Contributed by Alexander van
```
delete "having"

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -147314,83 +147758,180 @@ of an open range of nonnegative integers (of length equal to the length of
+
+  ${
+    $d U s $.  $d V s $.  $d W s $.
+    $( There exists a symbol such that its concatenation with the prefix
```
"its concatenation AFTER" is clearer: on first reading, I just thougth "how can this be true?" since I assumed "EARLIER" was meant.
The statement looks overly complex, but maybe it's used as a lemma in this form later.

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -148416,6 +149653,19 @@ Splicing words (substring replacement)
 
+  $c reverse $.
+
+  $( Syntax for the word reverse operator. $)
```
I think a better name for this operation is "word reversal"

avekens replied:
I did not create this definition, I just moved it to the corresponding subsection. Since this does not belong to the topic of this PR, I do not want to change this (without agreement of the contributor).

benjub commented on set.mm:
```diff
@@ -148416,6 +149653,19 @@ Splicing words (substring replacement)
+
+  ${
+    $d s x $.
+    $( Define an operation which reverses the order of symbols in a word.
+       (Contributed by Stefan O'Rear, 26-Aug-2015.) $)
```
Maybe add some synonyms, e.g., `This operation is also known as "word reversal" and "word mirroring".` This way, it will show up when someone searches for these expressions.

avekens replied:
Done.

benjub commented on set.mm:
```diff
@@ -147154,6 +147148,398 @@ of an open range of nonnegative integers (of length equal to the length of
+      UQUIURVFUQHVCABRCUJUKULUMVAURAUNUOUP $.
+
+    $d A x $.
+    $( Value of the prefix extractor as mapping.  (Contributed by AV,
```
I think it should be "as a mapping".

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -148551,6 +149801,20 @@ Splicing words (substring replacement)
+
+  ${
+    $d n s x $.
+    $( Definition to construct a word consisting of one repeated symbol, often
+       called "repeated symbol word" for short in the following.  (Contributed
+       by Alexander van der Vekens, 4-Nov-2018.) $)
+    df-reps $a |- repeatS
+                 = ( s e. _V , n e. NN0 |-> ( x e. ( 0 ..^ n ) |-> s ) ) $.
```
It looks more and more specialized. I hope there are (future?) applications to justify these definitions.

avekens replied:
This definition is not new, also only moved to the corresponding subsection. It is used to prove the friendship theorem.

benjub commented on set.mm:
```diff
@@ -150680,6 +152142,35 @@ the symbol at any position is repeated at multiples of L (modulo the
 
+  ${
+    $d W i $.
+    $( A prefix of length 2.  (Contributed by AV, 15-May-2020.) $)
```
2 --> two

avekens replied:
2 replaced by two.

benjub commented on set.mm:
```diff
@@ -150680,6 +152142,35 @@ the symbol at any position is repeated at multiples of L (modulo the
+      BXIXJXKXL $.
+  $}
+
   $( A word of length 3 represented as triple of ordered pairs.  (Contributed
```
3 --> three
(I may be a bit rigid... but to me, that's a bit like writing in all caps.)

avekens replied:
Done

benjub commented on set.mm:
```diff
@@ -150740,10 +152231,41 @@ the symbol at any position is repeated at multiples of L (modulo the
 
   $( Two words of length at least 2 are equal if and only if they have the same
```
2 --> two

avekens replied:
Done

benjub commented:
I added a few more comments, and I reviewed the whole thing (although not in depth). I'm not formally approving since there are a few comments by @savask and I above that require changes. But after they are resolved, it'll be ok for me.

avekens commented:
Thank you, @savask , for your remarks. I considered them in my latest commit.

> I looked at the changes. I don't have any serious remarks, besides perhaps the comment for `df-pfx`:
> 
> ```
>     $( Define an operation which extracts prefixes of words, i.e. subwords
>        starting at the beginning of a word.  Definition in Section 9.1 of
>        [AhoHopUll] p. 318.  "pfx" is used as label fragment as proposed in
>        ~ https://www.allacronyms.com/prefix/abbreviated .  (Contributed by AV,
>        2-May-2020.) $)
> ```
> 
> In my opinion, the reference to [AhoHopUll] is not required as the definition of a prefix is intuitive enough and, moreover, the definition was given at the start of the section `Prefixes of a word`. If you want to keep the reference it is much better to put it at the beginning of a section (replacing the stackexchange link, for example). I also suggest removing the reference to allacronyms.com completely, as it's not a credible source and we don't usually express any justification for using a specific abbreviation in set.mm.

The whole section header comment is removed now (as for the other subsections in this section) - the link to Wikipedia is moved to the comment for the definition ~df-pfx. 

To have the reference in the comment of the definition is a common practise in metamath, see, for example, df-iun: "Definition indexed union in [Stoll] p. 45.". I think also obvious/intuitive definitions should be referenced if possible.

The hint regarding the label fragment is also removed. All these information are still available in issue #1648.

> On the other hand it might be beneficial to hint the way this function works, for instance, like this:
> 
> ```
> Define an operation which extracts prefixes of words, i.e. subwords
> starting at the beginning of a word. In other words, ` ( S prefix L ) ` is
> a prefix of ` S ` of length ` L `.  (Contributed by AV, 2-May-2020.) 
> ```

I added this hint to the comment.

> On a more general note, I can't get rid of the feeling that it's a huge bulk of technical work without any immediate merit. Sure some theorems are shorter, but it would be great to see a nontrivial result involving prefixes (something like the "commuting strings theorem", i.e. `xy = yx` implies `x = z^n` and `y = z^m` for some string `z`. The proof uses prefixes, as far as I remember).

Everybody is invited to prove and contribute this theorem...

avekens commented on set.mm:
```diff
@@ -149815,6 +151265,18 @@ the symbol at any position is repeated at multiples of L (modulo the
       KYNWMXJEYNWLXIJKYNXIWLWJWPXIWLSXRWDUTVAVBVCWHVGVEVHVIWEBXBDEXSVNTWFWI $.
   $}
 
+  $( Mapping of words commutes with the prefix operation.  (Contributed by AV,
```
"mapping of words" is often used, I added the suggested explanation at its first occurence.

avekens commented:
I think all review remarks are considered now (as far as possible), so that this PR should be accepted and merged afterwards.

Thanks again, @benjub and @savask, for your valuable remarks.

benjub commented:
I'm getting the unicorn page more and more often lately (regardless of the PR, and even for small commits). Is it the same with you ?
When I can see the new commits, I'll probably approve.

avekens commented:
> I'm getting the unicorn page more and more often lately (regardless of the PR, and even for small commits). Is it the same with you ? When I can see the new commits, I'll probably approve.

What is the "unicorn page"? I have (and had) no problems with GitHub, except that the large complete diff is not shown (message "Oops, something went wrong.  Retry").

savask commented:
@benjub
>I'm getting the unicorn page more and more often lately (regardless of the PR, and even for small commits). Is it the same with you ?

I also haven't been able to check the new diff or even the raw set.mm file from the pull request page, but the new set.mm version is available in the relevant branch of avekens' fork.

benjub reviewed (approved):
Approved.  Thanks @avekens for the careful work.

benjub commented:
> What is the "unicorn page"? I have (and had) no problems with GitHub, except that the large complete diff is not shown (message "Oops, something went wrong. Retry").

 Right now, trying to access `https://github.com/metamath/set.mm/pull/2890/commits/4cc64b319ad5586d008a5a2f06bb52932b00ee77` I get what I call the "unicorn page" since there is a picture of an angry unicorn with the text:
```
This page is taking too long to load.

Sorry about that. Please try refreshing and contact us if the problem persists.
```
And at other times, I also get the message  "Oops, something went wrong. Retry".

But I checked on your branch as @savask suggested, so that's ok.
